node/cn/gasprice/gasprice.go

```diff
@@ -83,16 +91,40 @@ func NewOracle(backend OracleBackend, params Config, txPool TxPool) *Oracle {
 	if percent > 100 {
 		percent = 100
 	}
+
+	maxHeaderHistory := params.MaxHeaderHistory
+	if maxHeaderHistory < 1 {
+		maxHeaderHistory = 1
+		logger.Warn("Sanitizing invalid gasprice oracle max header history", "provided", params.MaxHeaderHistory, "updated", maxHeaderHistory)
+	}
+	maxBlockHistory := params.MaxBlockHistory
+	if maxBlockHistory < 1 {
+		maxBlockHistory = 1
+		logger.Warn("Sanitizing invalid gasprice oracle max block history", "provided", params.MaxBlockHistory, "updated", maxBlockHistory)
+	}
+	cache, _ := lru.New(2048)
+	headEvent := make(chan blockchain.ChainHeadEvent, 1)
+	backend.SubscribeChainHeadEvent(headEvent)
+	go func() {
+		var lastHead common.Hash
+		for ev := range headEvent {
+			if ev.Block.ParentHash() != lastHead {
+				cache.Purge()
+			}
+			lastHead = ev.Block.Hash()
+		}
+	}()
```

Comment:
If the subscription is for cache invalidation, we can simply listen to SetHead instead. I suspect that geth uses subscription because of the chain fork (a.k.a. [reorg](https://www.alchemy.com/overviews/what-is-a-reorg)) possibility (see [geth commit](https://github.com/ethereum/go-ethereum/pull/23422/commits/27ef4000cf13c9f85a569c088d20868ef56a08a8)). But Klaytn has no reorg, so it's sufficient to invalidate cache only when SetHead.

We can modify https://github.com/klaytn/klaytn/blob/dev/node/cn/api_backend.go
```go
func (b *CNAPIBackend) SetHead(number uint64) error {
  return doSetHead(b.cn.blockchain, b.cn.engine, b.cn.governance, b.gpo, number)
}

func doSetHead(..) {
  gpo.PurgeCache()
}
```

Reply:
Thanks for the suggestion, seems like it's much better approach.
Done in ba1ea0e.